dao/src/main/java/com/epam/esm/repository/TagRepositoryImpl.java

```diff
@@ -0,0 +1,64 @@
+package com.epam.esm.repository;
+
+import com.epam.esm.entities.Tag;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.RowMapper;
+
+import java.sql.ResultSet;
+import java.util.List;
+import java.util.Optional;
+
+@org.springframework.stereotype.Repository
+public class TagRepositoryImpl implements TagRepository {
+
+    public static final String GET_BY_ID = "SELECT * FROM tag WHERE id=?";
+    public static final String GET_ALL = "SELECT * FROM tag";
+    public static final String SAVE = "INSERT INTO tag (name) VALUE (LOWER(?)) ON DUPLICATE KEY UPDATE id=id;";
+    public static final String DELETE_BY_ID = "DELETE FROM tag WHERE id = ?";
+    private static final String SELECT_TAGS_BY_CERTIFICATE_ID = "SELECT t.name FROM tag t " +
+            "INNER JOIN tag_certificate tc on t.id = tc.tag_id " +
+            "INNER JOIN gift_certificate gc on tc.certificate_id = gc.id " +
+            "WHERE gc.id = ?";
+
+    private final JdbcTemplate jdbcTemplate;
+    private final RowMapper<Tag> mapper;
+
+    @Autowired
+    public TagRepositoryImpl(JdbcTemplate jdbcTemplate, RowMapper<Tag> mapper) {
+        this.jdbcTemplate = jdbcTemplate;
+        this.mapper = mapper;
+    }
+
+    @Override
+    public List<Tag> getAll() {
+        return jdbcTemplate.query(GET_ALL, mapper);
+    }
+
+    @Override
+    public void save(Tag entity) {
+        String name = entity.getName();
+        jdbcTemplate.update(SAVE, name);
+    }
+
+    @Override
+    public Optional<Tag> getById(int id) {
+        try {
+            return Optional.ofNullable(jdbcTemplate.queryForObject(GET_BY_ID, mapper, id));
+        } catch (EmptyResultDataAccessException e) {
+            return Optional.empty();
+        }
+    }
```

Comment:
I believe that you can avoid using try/catch in dao, just return Optional and check it in service layer

Reply:
queryForObject throws EmptyResultDataAccessException instead of null if query returns 0 rows. I don't know how to return Optional.empty() without catching the exception.

Reply:
Why you don`t use just a query() method? Is it deprecated?

Reply:
If i am not mistaken, you can use createQuery() or update() methods to execute delete as well

Reply:
> Why you don`t use just a query() method? Is it deprecated?

No, it't not depricated. I thought queryForObject fitted better in this case.